core/pkg/server/sender.go

```diff
@@ -226,11 +233,47 @@ func (s *Sender) Do(allWork <-chan runwork.Work) {
 		// TODO: reevaluate the logic here
 		s.configDebouncer.Debounce(s.upsertConfig)
 		s.summaryDebouncer.Debounce(s.streamSummary)
+
+		hangDetectionOutChan <- struct{}{}
 	}
+
+	close(hangDetectionInChan)
+	close(hangDetectionOutChan)
+
 	s.Close()
 	s.logger.Info("sender: closed", "stream_id", s.settings.RunId)
 }
 
+// warnOnLongOperations logs a warning for each message received
+// on the first channel for which no message is received on the second
+// channel within some time.
+func (s *Sender) warnOnLongOperations(
+	hangDetectionInChan <-chan runwork.Work,
+	hangDetectionOutChan <-chan struct{},
+) {
+outerLoop:
+	for work := range hangDetectionInChan {
+		start := time.Now()
+
+		for i := 0; ; i++ {
+			select {
+			case <-hangDetectionOutChan:
+				continue outerLoop
+
+			case <-time.After(30 * time.Minute):
+				if i < 6 {
+					s.logger.CaptureWarn(
+						"sender: taking a long time",
+						"seconds", time.Since(start).Seconds(),
+						"work", work.DebugInfo(),
```

Comment:
here you can probably use slog's nice feature to have special handle for work

Reply:
yeah, I think we can make `DebugInfo` the default `LogValue` implementation, but I'd like to think about that more since information-rich debug logging has been helpful in the past